perfkitbenchmarker/providers/ibmcloud/ibm_api.py

```diff
@@ -0,0 +1,577 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Utilities for working with IBM Cloud resources."""
+
+import threading
+import time
+import logging
+import json
+
+from absl import flags
+from perfkitbenchmarker.providers.ibmcloud import ibmcloud
+from perfkitbenchmarker.providers.ibmcloud import util
+
+FLAGS = flags.FLAGS
+
+AVAILABLE = 'available'
+RUNNING = 'running'
+STOPPED = 'stopped'
+
+
+class IbmAPICommand(object):
+  """A IBM Cloud rest api command.
+
+  Attributes:
+  
+    args: list of strings. Positional args to pass, typically
+        specifying an operation to perform (e.g. ['image', 'list'] to list
+        available images).
+        
+    flags: a dict mapping flag name string to flag value. Flags to pass to
+        ibm cloud rest api. If a provided value is True, the flag is passed without a value. 
+        If a provided value is a list, the flag is passed multiple
+        times, once with each value in the list.
+  """
+
+  gen = None
+  gen_instmgr = None
+  gen_imgmgr = None
+  gen_keymgr = None
+  gen_subnetmgr = None
+  gen_fipmgr = None
+  gen_volumemgr = None
+  gen_vpcmgr = None
+  gen_sgmgr = None
+  ibmcloud_account_id = None
+  ibmcloud_apikey = None
+  ibmcloud_auth_token = None
+  ibmcloud_auth_token_time = 0
+  _lock = threading.Lock()
+  base_time = 0.0
+  num_requests = None
+  debug = False
+
+  def __init__(self, resource, *args):
+    """Initializes a IbmAPICommand with the provided args and flags.
+
+    Args:
+      resource: A IBM Cloud resource of type BaseResource.
+      *args: sequence of strings. Positional args to pass to rest api calls,
+      typically specifying an operation to perform (e.g. ['image', 'list']
+      to list available images).
+    """
+    if args and args[0] == 'login':
+      IbmAPICommand.gen = ibmcloud.IbmCloud(account=self.ibmcloud_account_id,
+                                            apikey=self.ibmcloud_apikey,
+                                            verbose=self.debug,
+                                            version='v1',
+                                            silent=True,
+                                            force=False)
+      IbmAPICommand.gen_instmgr = ibmcloud.InstanceManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_imgmgr = ibmcloud.ImageManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_fipmgr = ibmcloud.FipManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_keymgr = ibmcloud.KeyManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_subnetmgr = ibmcloud.SubnetManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_volumemgr = ibmcloud.VolumeManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_vpcmgr = ibmcloud.VPCManager(IbmAPICommand.gen)
+      IbmAPICommand.gen_sgmgr = ibmcloud.SGManager(IbmAPICommand.gen)
+    self.args = list(args)
+    self.flags = {}
+    self.user_data = None
+    self.sgid = None
+
+  def GetToken(self):
```

Comment:
here's where you can use [vm_util.Retry](https://github.com/GoogleCloudPlatform/PerfKitBenchmarker/blob/master/perfkitbenchmarker/vm_util.py#L224) to retry this command if it fails

Reply:
yes, actually we've used it before, it helps sometimes, I will add.